climakitae/util/utils.py

```diff
@@ -200,45 +196,131 @@ def get_closest_gridcell(
         )
 
         # Convert coordinates to x,y
-        x, y = lat_lon_to_model_projection.transform(lon, lat)
+        # transform returns (x, y) with always_xy=True
+        coords1, coords2 = lat_lon_to_model_projection.transform(lon, lat)
+    else:
+        coords1 = lat
+        coords2 = lon
 
-    # Get closest gridcell using tolerance
-    # If input point outside of dataset by greater than one
-    # grid cell, then None is returned
-    try:
-        if "x" in data.dims and "y" in data.dims:
-            tolerance = km_num * 1000  # Converting km to m
-            closest_gridcell = data.sel(x=x, y=y, method="nearest", tolerance=tolerance)
-        elif "lat" in data.dims and "lon" in data.dims:
-            tolerance = km_num / 111  # Rough translation of km to degrees
-            closest_gridcell = data.sel(
-                lat=lat, lon=lon, method="nearest", tolerance=tolerance
+    dim1_idx = data[dim1_name].to_index().get_indexer([coords1], method="nearest")[0]
+    dim2_idx = data[dim2_name].to_index().get_indexer([coords2], method="nearest")[0]
+
+    if dim1_idx == -1 or dim2_idx == -1:
+        print("Input coordinate is OUTSIDE of data extent. Returning None.")
+        return None
+
+    # check for valid data at closest gridcell
+    # isel out all non spatial dimensions
+    test_data = data.isel({dim1_name: dim1_idx, dim2_name: dim2_idx})
+    for dim in data.dims:
+        if dim not in [dim1_name, dim2_name]:
+            test_data = test_data.isel({dim: 0})
+
+    # Check if data is valid (not all null)
+    # For Dataset, check if any data variable has valid data
+    # For DataArray, check if the data itself is valid
+    if isinstance(test_data, xr.Dataset):
+        has_valid_data = any(
+            not test_data[var].isnull().all() for var in test_data.data_vars
+        )
+    else:
+        has_valid_data = not test_data.isnull().all()
+
+    if has_valid_data:
+        closest_gridcell = data.isel({dim1_name: dim1_idx, dim2_name: dim2_idx})
+
+        if print_coords:
+            print(
+                f"Closest gridcell to lat: {lat}, lon: {lon} is at {dim1_name}: {closest_gridcell[dim1_name].item()}, {dim2_name}: {closest_gridcell[dim2_name].item()}"
```

Comment:
The function behavior is a bit strange in that it prints the wrong coordinate when indicating the closest gridcell. 

i.e.: 
```
Closest gridcell to lat: 38.507, lon: -121.496 is at x: -4065113.6618609736, y: 1408911.7306985362
```

It would be better to have it print the lat and lon coords.

Reply:
Fair enough. That's an easy enough fix.